File: chartered_book/web/static/recurring.js
/* The entries that come round again.

   Rent, salary, a loan instalment. The month one gets forgotten is the month
   the accounts are wrong, so what is due is shown rather than remembered.

   Nothing posts itself. A voucher that appeared without anybody agreeing to it
   is worse than one that was forgotten, because a forgotten one gets noticed
   and an invented one does not. What this removes is the remembering. */

var Recurring = (function () {
  "use strict";

  var el = UI.el, api = UI.api;

  App.register("recurring", function (page) {
    var box = el("div");

    function load() {
      return api("/api/recurring").then(draw);
    }

    function draw(data) {
      UI.clear(box);

      var owed = data.due_total;
      box.appendChild(el("div.card", {}, [
        el("div.card-head", {}, [
          el("h2", { text: owed
            ? (owed === 1 ? "One entry is due" : owed + " entries are due")
            : "Nothing is due" }),
          el("button.primary", { text: "Set up a new one",
                                 onclick: function () { editor(null); } })
        ]),
        el("p.card-note", { text: owed
          ? "Each one is posted on its own, so what goes into the books is seen "
            + "before it goes in."
          : "Rent, salary, a loan instalment. Set one up and it will be offered "
            + "each time it comes round, rather than needing to be remembered." })
      ]));

      (data.rows || []).forEach(function (pattern) {
        box.appendChild(card(pattern));
      });

      if (!(data.rows || []).length) {
        box.appendChild(el("div.empty", {}, [
          el("strong", { text: "Nothing set up yet" }),
          el("span", { text: "The entries that are the same every month belong here." })
        ]));
      }
    }

    function card(pattern) {
      var every = { month: "Every month", quarter: "Every quarter",
                    year: "Every year" }[pattern.every] || pattern.every;
      var head = el("div.card-head", {}, [
        el("h2", {}, [
          el("span", { text: pattern.name }),
          pattern.active ? null : el("span.pill.warn", { text: "switched off" })
        ]),
        el("div.row", {}, [
          el("button.secondary", { text: "Change",
                                   onclick: function () { editor(pattern); } }),
          App.state.permissions["voucher.cancel"]
            ? el("button.secondary", { text: "Remove", onclick: function () {
                UI.confirmAction("Remove " + pattern.name,
                  "The entries it has already made stay in the books. Those are real "
                  + "entries and have nothing to do with whether the pattern goes on.",
                  function () {
                    return api("/api/recurring/remove", { body: { id: pattern.id } })
                      .then(load)
                      .catch(function (error) { UI.flash(error.message, "bad"); });
                  }, "Remove it");
              }})
            : null
        ])
      ]);

      var lines = UI.table(
        ["Ledger", { label: "Debit", num: true }, { label: "Credit", num: true }],
        pattern.lines.map(function (line) {
          return el("tr", {}, [
            el("td", { text: line.account_code + "  " + line.account_name }),
            el("td.num", { text: line.dr_paisa ? UI.rs(line.dr_paisa) : "" }),
            el("td.num", { text: line.cr_paisa ? UI.rs(line.cr_paisa) : "" })
          ]);
        }), null, {});

      var due = el("div");
      if (pattern.due_count) {
        due.appendChild(el("div.doc-subtitle", { text: pattern.due_count === 1
          ? "One is due" : pattern.due_count + " are due" }));
        pattern.due.forEach(function (one) {
          due.appendChild(el("div.due-row", {}, [
            el("div", {}, [
              el("span.due-when", { text: UI.bs(one.date_ad, "long") }),
              el("span.muted", { text: "  " + one.date_ad })
            ]),
            el("button.secondary", { text: "Post this one", onclick: function (event) {
              var button = event.currentTarget;
              button.disabled = true;
              return api("/api/recurring/post",
                         { body: { id: pattern.id, due_bs: one.due_bs } })
                .then(function (result) {
                  UI.flash("Posted " + result.voucher.voucher.number + ".", "good");
                  return load();
                })
                .catch(function (error) {
                  button.disabled = false;
                  UI.flash(error.message, "bad");
                });
            }})
          ]));
        });
      } else if (pattern.active) {
        due.appendChild(el("p.card-note", { text: "Nothing due at the moment." }));
      }

      return el("div.card", {}, [
        head,
        el("p.card-note", { text: every + ", from "
          + UI.bs(bsToIso(pattern.starts_bs), "long")
          + (pattern.ends_bs ? " until " + UI.bs(bsToIso(pattern.ends_bs), "long") : "")
          + (pattern.narration ? "  ·  " + pattern.narration : "") }),
        lines,
        due
      ]);
    }

    /* Setting one up */

    function editor(existing) {
      var name = el("input", { type: "text", value: existing ? existing.name : "",
                               placeholder: "Shop rent" });
      var every = UI.select([
        { value: "month", label: "Every month" },
        { value: "quarter", label: "Every quarter" },
        { value: "year", label: "Every year" }
      ], existing ? existing.every : "month");
      var starts = UI.dateField(existing ? bsToIso(existing.starts_bs) : NP.todayIso());
      var ends = UI.dateField(existing && existing.ends_bs
                              ? bsToIso(existing.ends_bs) : "");
      var narration = el("input", { type: "text",
        value: existing ? existing.narration : "",
        placeholder: "What this entry is for" });

      var lines = (existing && existing.lines.length)
        ? existing.lines.map(function (line) {
            return { accountId: line.account_id, name: line.account_name,
                     dr: line.dr_paisa ? UI.rs(line.dr_paisa, { plain: true }) : "",
                     cr: line.cr_paisa ? UI.rs(line.cr_paisa, { plain: true }) : "" };
          })
        : [{ accountId: null, name: "", dr: "", cr: "" },
           { accountId: null, name: "", dr: "", cr: "" }];

      var grid = el("tbody");
      function drawLines() {
        UI.clear(grid);
        lines.forEach(function (line, index) {
          var account = el("input", { type: "text", value: line.name,
                                      placeholder: "Ledger" });
          UI.attachPicker(account, function (term) {
            return api("/api/accounts", { query: { q: term } })
              .then(function (d) { return d.rows; });
          }, function (chosen) {
            line.accountId = chosen.id;
            line.name = chosen.name;
            account.value = chosen.name;
          }, function (a) { return { main: a.name, side: a.group_name }; });

          var dr = UI.amountInput(line.dr, { onChange: function (v) { line.dr = v; } });
          var cr = UI.amountInput(line.cr, { onChange: function (v) { line.cr = v; } });
          dr.addEventListener("input", function () { line.dr = dr.value; });
          cr.addEventListener("input", function () { line.cr = cr.value; });

          grid.appendChild(el("tr", {}, [
            el("td", {}, [account]),
            el("td", {}, [dr]),
            el("td", {}, [cr]),
            el("td.no-print", {}, [el("button.link", { text: "×", onclick: function () {
              lines.splice(index, 1);
              if (lines.length < 2) { lines.push({ accountId: null, name: "", dr: "", cr: "" }); }
              drawLines();
            }})])
          ]));
        });
      }
      drawLines();

      var body = el("div", {}, [
        el("div.row", {}, [
          UI.field("Name", name),
          UI.field("How often", every)
        ]),
        el("div.row", {}, [
          UI.field("Starts", starts),
          UI.field("Ends, if it does", ends),
          UI.field("Narration", narration)
        ]),
        el("div.card-note", { text: "The same entry every time. A rent that changes is "
          + "a new pattern rather than a guess, so the amounts are fixed here." }),
        el("div.table-wrap", {}, [el("table", {}, [
          el("thead", {}, [el("tr", {}, [
            el("th", { text: "Ledger" }), el("th.num", { text: "Debit" }),
            el("th.num", { text: "Credit" }), el("th")
          ])]),
          grid
        ])]),
        el("button.secondary.no-print", { text: "Another line", onclick: function () {
          lines.push({ accountId: null, name: "", dr: "", cr: "" });
          drawLines();
        }})
      ]);

      UI.modal(existing ? "Change " + existing.name : "An entry that comes round again",
        body, [
        { label: "Cancel" },
        { label: existing ? "Save it" : "Set it up", kind: "primary", action: function () {
            var payload = {
              name: name.value, every: every.value,
              starts_bs: isoToBs(starts.getIso()),
              ends_bs: ends.getIso() ? isoToBs(ends.getIso()) : "",
              narration: narration.value, active: 1,
              lines: lines.filter(function (l) { return l.accountId; })
                          .map(function (l) {
                            return { account_id: l.accountId, dr: l.dr || 0, cr: l.cr || 0 };
                          })
            };
            var where = existing ? "/api/recurring/update" : "/api/recurring/create";
            if (existing) { payload.id = existing.id; }
            return api(where, { body: payload })
              .then(function () {
                UI.flash(existing ? "Saved." : "Set up.", "good");
                return load();
              })
              .catch(function (error) { UI.flash(error.message, "bad"); return false; });
          } }
      ], { wide: true });
    }

    // The engine keeps these dates as Bikram Sambat text; the date field works
    // in ordinary dates. One conversion in each direction, in one place.
    function bsToIso(bs) {
      if (!bs) { return ""; }
      var bits = String(bs).split("-");
      var day = +bits[2];
      // A pattern can be kept on a day some months do not have, so walk back
      // to one that exists rather than showing nothing.
      while (day > 27) {
        var iso = NP.bsToAd(+bits[0], +bits[1], day);
        if (iso) { return iso; }
        day -= 1;
      }
      return NP.bsToAd(+bits[0], +bits[1], day) || "";
    }
    function isoToBs(iso) {
      var bs = iso ? NP.adToBs(iso) : null;
      if (!bs) { return ""; }
      return bs.year + "-" + String(bs.month).padStart(2, "0")
             + "-" + String(bs.day).padStart(2, "0");
    }

    page.appendChild(box);
    return load();
  });

  return {};
}());

File: chartered_book/modules/recurring.py
"""
The entries that come round again.

Rent, salary, the electricity standing charge, a loan instalment, the monthly
depreciation charge. The same entry every month, typed again every month, and
the month it gets forgotten is the month the accounts are wrong.

Kept here as a pattern with the double entry it makes, so what is due can be
shown and posted rather than remembered.

Nothing posts itself, and that is deliberate. A voucher that appeared in the
books without anybody agreeing to it is worse than one that was forgotten,
because a forgotten one gets noticed and an invented one does not. What this
removes is the remembering, not the deciding.

Dates are Bikram Sambat throughout, because the months these things follow are
Nepali months. A rent due on the first of every month is due on the first of
Shrawan, not on the sixteenth of July.
"""

from ..core import audit, money, nepali_date as nd
from . import ledger

EVERY = {"month": 1, "quarter": 3, "year": 12}


class RecurringError(Exception):
    """Raised when a pattern cannot be made or run."""


def _now():
    from ..core import db
    return db.now_stamp()


def _parse_bs(text):
    """A Bikram Sambat date written 2083-04-01, as three numbers."""
    try:
        year, month, day = [int(part) for part in str(text).split("-")]
    except (ValueError, AttributeError):
        raise RecurringError("A date has to be written like 2083-04-01.")
    return year, month, day


def _format_bs(year, month, day):
    return "%04d-%02d-%02d" % (year, month, day)


def step_on(due_bs, every):
    """
    The next time after this one.

    A pattern due on the thirty second of a month has to survive the months
    that have thirty. It is pulled back to the last day of the month it lands
    in, and the day it was asked for is kept, so a run of months goes 32, 30,
    32 rather than being dragged down to 30 for ever by one short month.
    """
    months = EVERY.get(every)
    if not months:
        raise RecurringError("A pattern runs monthly, quarterly or yearly.")
    year, month, day = _parse_bs(due_bs)
    month += months
    while month > 12:
        month -= 12
        year += 1
    return _format_bs(year, month, day)


def _land_on(year, month, day):
    """Pull a day back to the last one the month actually has."""
    try:
        longest = nd.daysInMonth(year, month) if hasattr(nd, "daysInMonth") else None
    except Exception:                                               # noqa: BLE001
        longest = None
    if longest is None:
        longest = _days_in_month(year, month)
    return min(day, longest)


def _days_in_month(year, month):
    """How many days a Bikram Sambat month has, asked of the calendar itself."""
    for day in (32, 31, 30, 29):
        try:
            nd.bs_to_ad(year, month, day)
            return day
        except Exception:                                           # noqa: BLE001
            continue
    return 30


def to_ad(due_bs):
    """The ordinary date a Bikram Sambat due date falls on."""
    year, month, day = _parse_bs(due_bs)
    return nd.bs_to_ad(year, month, _land_on(year, month, day)).isoformat()


def create(conn, username, payload):
    """
    Set up one pattern.

    The double entry is checked here rather than when it runs, because a
    pattern that will not post is worse the twelfth time than the first.
    """
    name = (payload.get("name") or "").strip()
    if not name:
        raise RecurringError("Give it a name you will recognise in a list.")
    every = payload.get("every") or "month"
    if every not in EVERY:
        raise RecurringError("A pattern runs monthly, quarterly or yearly.")

    starts = (payload.get("starts_bs") or "").strip()
    if not starts:
        raise RecurringError("Say when it starts.")
    _parse_bs(starts)
    ends = (payload.get("ends_bs") or "").strip()
    if ends:
        _parse_bs(ends)
        if ends < starts:
            raise RecurringError("It cannot end before it starts.")

    lines = _clean_lines(conn, payload.get("lines") or [])

    cursor = conn.execute(
        """INSERT INTO recurring
           (name, voucher_type, every, day_of_month, starts_bs, ends_bs, next_due_bs,
            party_id, narration, note, active, created_by, created_at, updated_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, ?, ?, ?)""",
        (name[:120], payload.get("voucher_type") or "journal", every,
         _parse_bs(starts)[2], starts, ends, starts,
         payload.get("party_id") or None,
         (payload.get("narration") or "").strip()[:300],
         (payload.get("note") or "").strip()[:300],
         username, _now(), _now()))
    pattern_id = cursor.lastrowid
    _write_lines(conn, pattern_id, lines)
    audit.log(conn, username, "recurring.create", "recurring", pattern_id, name,
              "Set up %s, %s from %s" % (name, every, starts), None, None)
    return pattern_id


def update(conn, username, pattern_id, payload):
    """Change a pattern. What it has already posted is left alone."""
    row = conn.execute("SELECT * FROM recurring WHERE id = ?", (pattern_id,)).fetchone()
    if row is None:
        raise RecurringError("There is no such pattern.")
    lines = _clean_lines(conn, payload.get("lines") or [])
    every = payload.get("every") or row["every"]
    if every not in EVERY:
        raise RecurringError("A pattern runs monthly, quarterly or yearly.")
    ends = (payload.get("ends_bs") or "").strip()
    if ends:
        _parse_bs(ends)

    conn.execute(
        """UPDATE recurring SET name = ?, every = ?, ends_bs = ?, party_id = ?,
                  narration = ?, note = ?, active = ?, updated_at = ?
           WHERE id = ?""",
        ((payload.get("name") or row["name"]).strip()[:120], every, ends,
         payload.get("party_id") or None,
         (payload.get("narration") or "").strip()[:300],
         (payload.get("note") or "").strip()[:300],
         1 if payload.get("active", 1) else 0, _now(), pattern_id))
    conn.execute("DELETE FROM recurring_lines WHERE recurring_id = ?", (pattern_id,))
    _write_lines(conn, pattern_id, lines)
    audit.log(conn, username, "recurring.update", "recurring", pattern_id,
              row["name"], "Changed %s" % row["name"], None, None)
    return True


def _clean_lines(conn, raw):
    """The double entry, checked now rather than every time it runs."""
    lines = []
    total_dr = total_cr = 0
    for index, row in enumerate(raw, start=1):
        account_id = row.get("account_id")
        if not account_id:
            continue
        account = conn.execute("SELECT id, name FROM accounts WHERE id = ?",
                               (account_id,)).fetchone()
        if account is None:
            raise RecurringError("Line %d points at a ledger that is not there." % index)
        dr = money.to_paisa(row.get("dr") or 0)
        cr = money.to_paisa(row.get("cr") or 0)
        if dr and cr:
            raise RecurringError("Line %d is on both sides at once." % index)
        if dr < 0 or cr < 0:
            raise RecurringError("Line %d cannot be a negative amount." % index)
        if not dr and not cr:
            continue
        total_dr += dr
        total_cr += cr
        lines.append({"account_id": account_id, "dr": dr, "cr": cr,
                      "narration": (row.get("narration") or "").strip()[:200]})
    if len(lines) < 2:
        raise RecurringError("An entry needs at least two lines.")
    if total_dr != total_cr:
        raise RecurringError(
            "It does not balance. Debits come to %s and credits to %s."
            % (money.format_money(total_dr), money.format_money(total_cr)))
    return lines


def _write_lines(conn, pattern_id, lines):
    conn.executemany(
        """INSERT INTO recurring_lines
           (recurring_id, line_no, account_id, dr_paisa, cr_paisa, narration)
           VALUES (?, ?, ?, ?, ?, ?)""",
        [(pattern_id, n, line["account_id"], line["dr"], line["cr"], line["narration"])
         for n, line in enumerate(lines, start=1)])


def listing(conn, as_at_bs=None):
    """
    Every pattern, with what it owes and what it has already made.

    The number due is the point of the screen, so it is worked out here rather
    than left for the screen to count.
    """
    as_at_bs = as_at_bs or _format_bs(*nd.today_bs())
    out = []
    for row in conn.execute("SELECT * FROM recurring ORDER BY active DESC, name"):
        pattern = dict(row)
        pattern["lines"] = [dict(line) for line in conn.execute(
            """SELECT l.*, a.code AS account_code, a.name AS account_name
               FROM recurring_lines l JOIN accounts a ON a.id = l.account_id
               WHERE l.recurring_id = ? ORDER BY l.line_no""", (row["id"],))]
        pattern["amount"] = sum(line["dr_paisa"] for line in pattern["lines"])
        pattern["due"] = due_list(conn, row["id"], as_at_bs)
        pattern["due_count"] = len(pattern["due"])
        pattern["posted_count"] = conn.execute(
            "SELECT COUNT(*) AS n FROM recurring_posted WHERE recurring_id = ?",
            (row["id"],)).fetchone()["n"]
        out.append(pattern)
    return {"rows": out, "as_at_bs": as_at_bs,
            "due_total": sum(p["due_count"] for p in out)}


def due_list(conn, pattern_id, as_at_bs=None):
    """
    Every date this pattern owes up to today, oldest first.

    Every one, not only the next. A pattern set up in Shrawan and looked at in
    Mangsir owes four months, and showing one of them would leave three quietly
    missing, which is exactly the failure this is meant to end.
    """
    as_at_bs = as_at_bs or _format_bs(*nd.today_bs())
    row = conn.execute("SELECT * FROM recurring WHERE id = ?", (pattern_id,)).fetchone()
    if row is None or not row["active"]:
        return []

    already = {done["due_bs"] for done in conn.execute(
        "SELECT due_bs FROM recurring_posted WHERE recurring_id = ?", (pattern_id,))}

    due = []
    when = row["starts_bs"]
    # A guard on the loop rather than trust in the dates: a pattern with a
    # start far in the past should give a long list, not run for ever.
    for _ in range(400):
        if when > as_at_bs:
            break
        if row["ends_bs"] and when > row["ends_bs"]:
            break
        if when not in already:
            year, month, day = _parse_bs(when)
            landed = _format_bs(year, month, _land_on(year, month, day))
            due.append({"due_bs": landed, "date_ad": to_ad(when)})
        when = step_on(when, row["every"])
    return due


def post_due(conn, username, pattern_id, due_bs):
    """
    Make the entry for one date.

    Refused where that date has already been made, which is what stops a month
    being posted twice by two people looking at the same screen.
    """
    row = conn.execute("SELECT * FROM recurring WHERE id = ?", (pattern_id,)).fetchone()
    if row is None:
        raise RecurringError("There is no such pattern.")
    if not row["active"]:
        raise RecurringError("That pattern has been switched off.")

    done = conn.execute(
        "SELECT voucher_id FROM recurring_posted WHERE recurring_id = ? AND due_bs = ?",
        (pattern_id, due_bs)).fetchone()
    if done is not None:
        raise RecurringError("That one has already been posted.")

    lines = conn.execute(
        "SELECT * FROM recurring_lines WHERE recurring_id = ? ORDER BY line_no",
        (pattern_id,)).fetchall()
    if not lines:
        raise RecurringError("That pattern has no entry to make.")

    voucher_id = ledger.post_voucher(conn, username, {
        "voucher_type": row["voucher_type"] or "journal",
        "date_ad": to_ad(due_bs),
        "party_id": row["party_id"],
        "narration": row["narration"] or row["name"],
        "entries": [{"account_id": line["account_id"],
                     "dr": money.to_rupees(line["dr_paisa"]),
                     "cr": money.to_rupees(line["cr_paisa"]),
                     "narration": line["narration"]} for line in lines],
    })

    conn.execute(
        """INSERT INTO recurring_posted (recurring_id, due_bs, voucher_id, posted_by, posted_at)
           VALUES (?, ?, ?, ?, ?)""",
        (pattern_id, due_bs, voucher_id, username, _now()))
    conn.execute("UPDATE recurring SET next_due_bs = ?, updated_at = ? WHERE id = ?",
                 (step_on(due_bs, row["every"]), _now(), pattern_id))
    audit.log(conn, username, "recurring.post", "recurring", pattern_id, row["name"],
              "Posted %s for %s" % (row["name"], due_bs), None, None)
    return voucher_id


def remove(conn, username, pattern_id):
    """
    Take a pattern away.

    What it has already posted stays. Those are real entries in the books and
    have nothing to do with whether the pattern goes on running.
    """
    row = conn.execute("SELECT name FROM recurring WHERE id = ?", (pattern_id,)).fetchone()
    if row is None:
        raise RecurringError("There is no such pattern.")
    conn.execute("DELETE FROM recurring WHERE id = ?", (pattern_id,))
    audit.log(conn, username, "recurring.remove", "recurring", pattern_id, row["name"],
              "Removed the pattern %s. Entries already posted are untouched."
              % row["name"], None, None)
    return True
